### PR TITLE
feature: Separated Add Resources to checklist

### LINKED ISSUE #6095: Check boxes not showing next to be installed items in Required
## Describe the bug

Check boxes in Resource selection screen for publishing profile should indicate that it will be installed.  Required resources should show as checked and you should not be able to unselect them.

## Version

image

## Browser

- [X] Electron distribution
- [ ] Chrome
- [ ] Safari
- [ ] Firefox
- [ ] Edge

## OS

- [X] macOS
- [X] Windows
- [ ] Ubuntu

## To Reproduce

1. Create a new Publishing Profile and get to Add Resources page
2. Check boxes for required resources not checked and is confusing as it looks like nothing will be installed if you uncheck optional items

## Expected behavior

Things that are going to be installed should be indicated in the UI for Required items.  Otherwise it is very confusing what will and will not be installed.  Check boxes should be checked next to resources but are in not uncheckable.

## Screenshots

image

### LINKED ISSUE #6299: Add Resources deselects all selected resources when I click on the resource name
## Describe the bug

When creating publish profile there are checkboxes next to the optional resources.  If I click on the checkboxes it toggles the check for the item I click on.  But I click on the resource name, it deselects ALL checkboxes.  That seems wrong.

## Version

1.3.1-nightly20225

## Browser

- [x] Electron distribution
- [ ] Chrome
- [ ] Safari
- [ ] Firefox
- [ ] Edge

## OS

- [ ] macOS
- [x] Windows
- [ ] Ubuntu

## To Reproduce

 See screenshot 

## Expected behavior

Toggling each one seperately seems like desired behavior.

## Screenshots

image

### PR DESCRIPTION
**Pull request by GeoffCoxMSFT**
## Description

AddResources step of details page used DetailsList which had interaction problems with required resources (not displaying checkboxes) and with readability/layout.  

This replaces that step with a checklist style to solve the missing checkbox problem and improve the experience.

## Task Item

fixes #6299 (fixed by removing DetailsList)
fixes #6095 

## Screenshots
image
image

### CODE REVIEW

File: extensions/azurePublish/src/components/ChooseResourcesList.tsx [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+`;
+
+const ItemTier = styled(Text)`
+  font-size: 12px;
+  margin: 4px 0 0 22px;
+  color: ${NeutralColors.gray130};
+`;
+
+const ItemDescription = styled(Text)`
+  font-size: 14px;
```
Comment (hatpick) [inline extensions/azurePublish/src/components/ChooseResourcesList.tsx:60]: use FluentTheme.fonts

File: extensions/azurePublish/src/components/ChooseResourcesList.tsx [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+ * Raises a callback of the selected keys.
+ * Allows for uncontrolled or controlled selected keys.
+ */
+export const ChooseResourcesList = (props: Props) => {
+  const { items, selectedKeys: controlledSelectedKeys, onSelectionChanged } = props;
+
+  // ----- Hooks
+
+  const [selectedKeys, setSelectedKeys] = React.useState<string[]>(() => {
+    return controlledSelectedKeys || items.filter((item) => item.required).map((item) => item.key);
```
Comment (hatpick) [inline extensions/azurePublish/src/components/ChooseResourcesList.tsx:97]: extract this into a util function to be reused in L102

File: extensions/azurePublish/src/components/ChooseResourcesList.tsx [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+  React.useEffect(() => {
+    setSelectedKeys(controlledSelectedKeys || items.filter((item) => item.required).map((item) => item.key));
+  }, [items, controlledSelectedKeys]);
+
+  // ----- Handlers
+
+  const onCheckboxChanged = (ev: React.FormEvent<HTMLElement>, checked: boolean, item: ResourceListItem): void => {
+    let newSelectedKeys = undefined;
+    if (item.required || checked) {
+      if (!selectedKeys.find((i) => i === item.key)) {
```
Comment (hatpick) [inline extensions/azurePublish/src/components/ChooseResourcesList.tsx:110]: nit: .includes(item.key)

File: extensions/azurePublish/src/components/ChooseResourcesList.tsx [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+          <ItemText>{item.text}</ItemText>
+        </ItemHeader>
+        <ItemTier>{item.tier}</ItemTier>
+        <ItemDescription>{item.description}</ItemDescription>
+      </ItemLabel>
+    );
+  };
+
+  const renderItem = (item: ResourceListItem) => {
+    const checked = item.required || !!selectedKeys.find((selectedKey) => selectedKey === item.key);
```
Comment (hatpick) [inline extensions/azurePublish/src/components/ChooseResourcesList.tsx:140]: .includes() returns boolean

File: extensions/azurePublish/src/components/ChooseResourcesList.tsx [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        disabled={item.required}
+        styles={checkBoxStyle}
+        onChange={(e, c) => onCheckboxChanged(e, c, item)}
+        onRenderLabel={() => renderItemLabel(item)}
+      />
+    );
+  };
+
+  return (
+    <div>
```
Comment (hatpick) [inline extensions/azurePublish/src/components/ChooseResourcesList.tsx:155]: why do you need the wrapper div?

File: extensions/azurePublish/src/components/azureProvisionDialog.tsx [hunk 6]
```diff
@@ -38,12 +40,18 @@ import { AzureResourceTypes, ResourcesItem, authConfig } from '../types';
...
   getPreview,
   getLuisAuthoringRegions,
   CheckWebAppNameAvailability,
 } from './api';
+import { ChooseResourcesList } from './ChooseResourcesList';
+
+// ---------- Styles ---------- //
+
+const AddResourcesSectionName = styled(Text)`
+  font-size: 16px;
```
Comment (hatpick) [inline extensions/azurePublish/src/components/azureProvisionDialog.tsx:53]: use FluentTheme.fonts

File: extensions/azurePublish/src/components/azureProvisionDialog.tsx [hunk 7]
```diff
@@ -334,8 +295,7 @@ export const AzureProvisionDialog: React.FC = () => {
   const [importConfig, setImportConfig] = useState<any>();
 
   const [page, setPage] = useState(PageTypes.ConfigProvision);
-  const [group, setGroup] = useState<IGroup[]>();
-  const [listItems, setListItem] = useState<(ResourcesItem & { name; icon })[]>();
+  const [listItems, setListItem] = useState<(ResourcesItem & { icon?: string })[]>();
```
Comment (hatpick) [inline extensions/azurePublish/src/components/azureProvisionDialog.tsx:298]: do you want the default value be undefined or an empty list?

File: extensions/azurePublish/src/components/azureProvisionDialog.tsx [hunk 8]
```diff
@@ -410,6 +363,13 @@ export const AzureProvisionDialog: React.FC = () => {
     }
   };
 
+  useEffect(() => {
+    if (token) {
+      getSubscriptions(token).then(setSubscriptions);
```
Comment (hatpick) [inline extensions/azurePublish/src/components/azureProvisionDialog.tsx:368]: make sure the component is not unmounted when the setState is called.

File: extensions/azurePublish/src/components/azureProvisionDialog.tsx [hunk 9]
```diff
@@ -747,43 +692,39 @@ export const AzureProvisionDialog: React.FC = () => {
...
-        return item.required === false;
-      },
-    });
-    if (s && listItems) {
-      s.setItems(listItems, false);
-      s.setAllSelected(true);
-    }
-    return s;
-  }, [listItems]);
+  const PageAddResources = () => {
```
Comment (hatpick) [inline extensions/azurePublish/src/components/azureProvisionDialog.tsx:695]: this is like a component defined within another component. why not extract it to it's own component?

File: extensions/azurePublish/src/components/azureProvisionDialog.tsx [hunk 10]
```diff
@@ -747,43 +692,39 @@ export const AzureProvisionDialog: React.FC = () => {
...
+          <Stack>
+            <AddResourcesSectionName>{formatMessage('Required')}</AddResourcesSectionName>
+            <ChooseResourcesList items={requiredListItems} />
+            <AddResourcesSectionName>{formatMessage('Optional')}</AddResourcesSectionName>
+            <ChooseResourcesList
+              items={optionalListItems}
+              selectedKeys={selectedResourceKeys}
+              onSelectionChanged={(keys) => {
+                const newSelection = listItems.filter(
+                  (item) => item.required === true || keys.find((key) => key === item.key)
```
Comment (hatpick) [inline extensions/azurePublish/src/components/azureProvisionDialog.tsx:716]: replace .find with .includes or .some

File: extensions/azurePublish/src/components/azureProvisionDialog.tsx [hunk 11]
```diff
@@ -747,43 +692,39 @@ export const AzureProvisionDialog: React.FC = () => {
...
-            getKey={(item) => item.key}
-            groups={group}
-            items={listItems}
-            layoutMode={DetailsListLayoutMode.justified}
-            selection={selection}
-            selectionMode={SelectionMode.multiple}
-            setKey="none"
-          />
+      return (
+        <ScrollablePane
```
Comment (hatpick) [inline extensions/azurePublish/src/components/azureProvisionDialog.tsx:702]: this control needs it's parent to be set to position: relative, is that the case?

File: extensions/azurePublish/src/components/ChooseResourcesList.tsx [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+`;
+
+const checkBoxStyle = {
+  label: {
+    alignItems: 'flex-start',
+  },
+};
+
+const ItemLabel = styled(Stack)`
+  margin-left: 15px !important;
```
[the remark this answers is not in the document]
Comment (GeoffCoxMSFT) [reply extensions/azurePublish/src/components/ChooseResourcesList.tsx:29]: Stack style overrides margin-left (0px) and won't accept the styled component without !important.

File: extensions/azurePublish/src/components/azureProvisionDialog.tsx (hunk 9 above)
Comment (GeoffCoxMSFT) [reply extensions/azurePublish/src/components/azureProvisionDialog.tsx:677]: That will happen eventually.

File: extensions/azurePublish/src/components/ChooseResourcesList.tsx [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+const ImageIconPlacholder = styled.div`
+  width: 16px;
+  height: 16px;
+  margin: 4px 0 0 0;
+  user-select: none;
+`;
+
+const ItemText = styled(Text)`
+  font-size: 16px;
+  margin-left: 4px !important;
```
[the remark this answers is not in the document]
Comment (GeoffCoxMSFT) [reply extensions/azurePublish/src/components/ChooseResourcesList.tsx:50]: Same reason as before.  child[0] gets a margin-left: 0px set by fluent.

File: extensions/azurePublish/src/components/azureProvisionDialog.tsx (hunk 7 above)
Comment (GeoffCoxMSFT) [reply extensions/azurePublish/src/components/azureProvisionDialog.tsx:245]: I'm removing the name from the type as it isn't needed anywhere.  I'm not changing the default value.

File: extensions/azurePublish/src/components/azureProvisionDialog.tsx (hunk 8 above)
Comment (GeoffCoxMSFT) [reply extensions/azurePublish/src/components/azureProvisionDialog.tsx:343]: @hatpick How do I do that with hooks? (I pulled this code directly from another existing component)

File: extensions/azurePublish/src/components/azureProvisionDialog.tsx (hunk 11 above)
Comment (GeoffCoxMSFT) [reply extensions/azurePublish/src/components/azureProvisionDialog.tsx:684]: I don't think so, the calc thing was there before and makes the control move up from the bottom where the buttons are on the wizard.  The parent wizard should have used grid or stack so children don't have to do this calc.

**Review by hatpick — DISMISSED** [review]

**Review by benbrown — DISMISSED** [review]
Code looks awesome.

Small changes requested:
* Default optional to on.  This is the current behavior we should preserve til we can discuss with PMs.
* Hide the required header if the required list is empty.
* Verify we didn't regress "edit" scenario where there are no remaining resources available (should jump to JSON view)
* Allow design to take a look and give final comments

**Comment by GeoffCoxMSFT** [comment]
* Default optional to on.  This is the current behavior we should preserve til we can discuss with PMs.
> DONE
* Hide the required header if the required list is empty.
> DONE
* Verify we didn't regress "edit" scenario where there are no remaining resources available (should jump to JSON view)
> No change to that code path.  Can't get there as my subscription won't successfully deploy all the optional parts. 
* Allow design to take a look and give final comments
> UX can inspect the nightlies and give feedback

**Pull request merged**